docs/proposals/custom-metrics-api.md

```diff
@@ -0,0 +1,189 @@
+  given name in the given namespace
+
+For example, to retrieve the custom metric "hits-per-second" for all pods
+matching "app=frontend` in the namespaces "webapp", the request might look
+like:
+
+`/apis/general-metrics/v1alpha1/namespaces/webapp/object-metrics/pods/hits-per-second?labelSelector=app%3Dfrontend`.
+
+API Objects
+-----------
+
+The request URLs listed above will return objects of the `Metrics`
+type, described below:
+
+```go
+
+// a list of values for a given metric for some set of objects
+type Metrics struct {
+    unversioned.TypeMeta `json:",inline"`
+    unversioned.ListMeta `json:"metadata,omitempty"`
+
+    // the name of the metric
+    MetricName string `json:"metricName"`
+
+    // the value of the metric across the described objects
+    MetricValues []MetricValue `json:"metricValues"`
+}
+
+// a metric value for some object
+type GenericMetrics struct {
+    // the type and object metadata describes the object in question
+    unversioned.TypeMeta `json:",inline"`
+    v1.ObjectMeta        `json:"metadata,omitempty"`
+
+    // indicates the end of the time window containing these metrics (i.e.
+    // these metrics come from some time in [Timestamp-Window, Timestamp])
+    Timestamp unversioned.Time `json:"timestamp"`
+
+    // indicates the duration of the time window containing these metrics
+    Window    unversioned.Duration `json:"window"`
+
+    // the value of the metric for this
+    Value float64
+}
+```
+
+For instance, the example request above would yield the following object:
+
+```json
+{
+    "kind": "Metrics",
+    "apiVersion": "general-metrics/v1alpha1",
+    "metricName": "hits-per-second",
+    "metricValues": [
+        {
+            "kind": "Pod",
+            "metadata": { "name": "server1", "namespace": "webapp" },
+            "timestamp": SOME_TIMESTAMP_HERE,
+            "window": "10s",
+            "value": 10.0
+        },
+        {
+            "kind": "Pod",
+            "metadata": { "name": "server2", "namespace": "webapp" },
+            "timestamp": SOME_TIMESTAMP_HERE,
+            "window": "10s",
+            "value": 15.0
+        }
+    ]
+}
+```
+
+Mechanical Concerns
+-------------------
+
+This API is intended to be implemented by monitoring pipelines (e.g.
+inside Heapster, or as an adapter on top of a solution like Prometheus).
+It shares many mechanical requirements with normal Kubernetes APIs.
+
+The API should support encoding the result in both JSON and Protobuf,
+similarly to how core Kubernetes APIs work.  Most likely, the Protobuf
+encoding will be used by most consumers for efficiency's sake.
+
+Additionally, the API should support encoding the output result in
+different versions, if needed, as is currently the case for normal
+Kubernetes APIs.
+
+### Discoverability ###
+
+While the initial plan for the API is for consumption by the HPA
+controller, the API could be more broadly useful -- custom autoscaler or
```

Comment:
I know at the moment we have users who have written applications against the Heapster model API, so I think we should keep in mind that others might want to use the API.

That being said, I agree that it shouldn't be too generic.  The only thing in this API that the HPA v2 can't use is selecting multiple objects that aren't pods, but preventing this just makes the API more complicated.